cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,277 @@
+
+### Label sizes
+
+CIDR labels, especially IPv6 labels, are large. For example, the full set of labels for the CIDR `2001:db8:cafe::cab:4:b0b:0/112` is 2KiB. Every extra-cluster identity always has the full set of CIDR identities calculated for it.
+
+----
+
+## Proposal
+
+This CFP proposes 3 key changes. Together, they preserve correctness while radically restructuring how prefixes are selected.
+
+1: Labels on prefixes propagate downwards. That is to say, if prefix `192.0.0.0/22` has label `foo:bar`, then prefix `192.0.2.0/24` also has label `foo:bar`, unless overridden. The IPCache aggregates labels accordingly.
+2: Only selected CIDR labels are inserted in to the ipcache. If a selector references CIDR `192.0.2.0/24`, only the prefix -> label mapping `192.0.2.0/24 -> cidr:192.0.2.0/24` is added to the IPCache.
+3: When the NameManager learns about a new relevant `name -> IP` pair, it upserts the label `IP -> (dns:selector)` It **does not** create any CIDR labels.
+
+
+### 1: CIDR Label down-propagation
+
+Currently, prefixes in the ipcache are independent. Instead, labels should flow downwards from prefixes. For example, consider an IPCache with two prefixes: 
+
+```
+192.0.2.0/24 -> k8s:foo:bar, cidr:192.0.2.0/24
+192.0.2.2/32 -> cidr:192.0.2.2/32, reserved:kube-apiserver
+```
+
+Then, the complete set of labels for `192.0.2.2/32` would be `k8s:foo:bar, cidr:192.0.2.0/24, cidr:192.0.2.2/32, reserved:kube-apiserver`.
+
+This is a cornerstone for implementing another feature, non-k8s label sources. It is also required for correctness within the context of this proposal, especially when ToFQDN and CIDR selectors overlap.
+
+#### Lifecycle
+
+With label down-propagation, it is important that updates are propagated accordingly. When a prefix P is updated or deleted, then updates must be also triggered for all prefixes contained in P.
+
+This update may be calculated if we use a bitwise LPM trie. Alternatively, scanning the entire ipcache to determine affected prefixes may be more efficient.
+
+
+### 2: Only generate selected CIDR labels
+
+Currently, every selected prefix gets the full set of CIDR labels. In other words, the prefix `192.168.0.42/32` will have 32 labels generated. However, this is not truly necessary. The key insight is that the ipcache already knows about all selected CIDRs, since selectors are required to upsert relevant CIDRs with the IPCache before referencing them.
+
+So, if selector A selects `cidr:192.168.0.0/24`, there will be a single identity, `id1 -> (cidr:192.168.0.0/24)`. If selector B selects `cidr:192.168.0.0/25`, then there will be two identities: `id1-> (cidr:192.168.0.0/24)` and `id2 -> (cidr:192.168.0.0/25, cidr:192.168.0.0/24)`. Selector A will now select _id1_ and _id2_, and selector B will select just _id2_.
```

Comment:
The downside here is potential to increase the amount of dynamic state recalculation when a new policy is created.

I think we could alternatively explore an option where we avoid encoding each prefix's labels in the Identities and instead have an optional CIDR association in the local Identities. Then extend the label selector so that when it evaluates selections against an Identity with an embedded CIDR, it can do a CIDR subset check using straight math rather than label comparisons.

Reply:
> The downside here is potential to increase the amount of dynamic state recalculation when a new policy is created.

Right, I had a section musing about this, but it never went anywhere and I axed it. Because of FQDN policies, we need to have a dynamic policy engine. I think it's important to _match impedances_, that is to say, try and have more frequent and latency-sensitive operations be less expensive, and infrequent operations be more expensive. 

Right now, ingesting policy and adding new IPs to an FQDN selector are near equivalent to the SelectorCache and IPCache. However, these take place in totally different contexts; one happens very occasionally and has latency measured in seconds (k8s strives for a pod creation time of ~2 seconds), and the other happens constantly and has a hard deadline of 100ms.

From this perspective, it makes sense to shift as much computation as possible to the slow, latent path. We can't take this too far, however, since the occasional slow operation still blocks any fast ones that happen to be waiting in line.

I am a bit concerned that a pathological sequence of ToCIDR policy updates could cause a surge of load. On the whole, however, I think this would be a huge improvement in experience. We essentially have that pathological sequence of updates now with S3, so it's unlikely to be any worse ;-).

Reply:
I agree in principle with that path. I guess the idea I'm broaching in the counter-proposal above is whether we need to internally encode the IP Prefixes as strings at all. Probably for external interfaces like when we serialize flows on the wire we would need to include string representations of the CIDR labels, but maybe we don't have to internally track IP Prefixes in strings at all (if we just make the selectors, ipcache, etc a bit smarter wrt prefix handling and encoding the prefixes using netip.Prefix).

Reply:
Thats a really interesting avenue of exploration. It does allow for short-cutting certain cases in this proposal that would require re-allocation.

I wonder if this is basically equivalent to having mutable labels on local identities. Have to stare at the wall for a bit and decide.

Reply:
My initial inclination is that mutable labels is a separate, more complex problem. We could start by stating that if we must encode an IP prefix into an Identity, there would only be one prefix and it would not change. If the prefix needs to change, we would allocate a new Identity. This is closer to what we have today.

Reply:
Oooh I wonder if we can make this even easier: make the `cidr:xxx` label selector smarter and match longer contained prefixes.

Reply:
> make the cidr:xxx label selector smarter and match longer contained prefixes.

Ah right, I remember why I didn't propose this: the lifecycle gets ugly with multiple overlapping ToFQDN and CIDR selectors. Label down-propagation solves so many problems so neatly.

Reply:
Ah. I guess part of this is if we try to visualize the ideal representation of the following cases:

- Only ToFQDN rules exist
- There is a ToFQDN rule and a specific CIDR rule or multiple CIDR rules
- There is a ToFQDN rule and a broad CIDR rule

Then work backwards from those "steady states" to figure out how we might represent these in the ipcache, identity and selector... and work through the lifecycles of how to transition into these states (eg starting from no policy -> tofqdn, then introduce CIDR policy, remove CIDR policy etc.).